<a id="1"></a>
<div style="padding:20px;
            color:#F0F3FF;
            margin:10;
            font-size:200%;
            text-align:left;
            display:fill;
            border-radius:15px;
            background-color:#211951;
            overflow:hidden;
            font-weight:900">
            <span style='color:#15F5BA'></span>Pandas : Data Manipulation
</div>
<br>Author : <i>Retno Sari Sirotujanah</i>

# <b>A. <span style='color:#3A6D8C'>Introduction</span></b>

Sebelum sebuah data dapat dianalisis atau digunakan untuk mengambil keputusan kebijakan tentu data tersebut perlu untuk diekstraksi terlebih dahulu. <b>Ekstraksi data</b> adalah sebuah proses pengumpulan atau pengambilan jenis data yang berbeda dari berbagai sumber, banyak di antaranya mungkin tidak terorganisir dengan baik atau tidak terstruktur sama sekali. Proses ini memungkinkan untuk mengkonsolidasikan, memproses, dan menyempurnakan data sehingga nantinya dapat disimpan di lokasi terpusat untuk diubah atau diproses lebih lanjut.

Python sebagai salah satu <i>tools</i> yang <i>powerful</i> untuk pengolahan data menyediakan berbagai <i>library</i> untuk mendukung proses ekstraksi data. Salah satu library paling banyak digunakan adalah <b>pandas</b>.

Pandas memungkinkan untuk membaca data dari berbagai jenis ekstensi seperti data yang bersumber dari database (sql), <i>excel</i> (<b>.xls</b> atau <b>.xlsx</b>), <i>separated value file</i> (<b>.csv</b>, <b>.tsv</b>, dsb) lalu merubahnya menjadi satu format struktur data yang sama yang disebut <b>DataFrame</b>. DataFrame inilah yang nantinya dapat kita lakukan proses <b>Data Manipulation</b><br><br>

<p align="center">
<img src="https://raw.githubusercontent.com/bachtiyarma/Material/refs/heads/main/Image/Materi-Python/Pandas-DataFrame.png" width="600" height="300">
<p align="center"><b>Gambar 1. Proses Ekstraksi Data dengan <i>pandas</i></b></p>
<br><br>
<p align="left"><sup>docs : <i><a href="https://shorturl.at/xyZ69">https://pandas.pydata.org/docs/user_guide/io.html</i></sup></a></p>


## <b>A.1. <span style='color:#6A9AB0'>First Step : Pandas</span></b>  
Untuk dapat menggunakan library pandas pada python, perlu untuk mengimpor library tersebut dengan cara mengetikan sintaks berikut

<p align="center"><code><b>import pandas as pd</b></code></p>

lalu jalankan (<i>run</i>) program.
<br><br>
<p align="left"><sup>docs : <i><a href="https://shorturl.at/kzG06">https://pandas.pydata.org/docs/getting_started/install.html</i></sup></a></p>

###<b><span style='color:#55679C'>Quest A.1</span></b>
<i>Ketikan kode `import pandas as pd` pada prompt dibawah ini lalu jalankan!</i>

In [ ]:
# Answer A.1
import pandas as pd

# <b>B. <span style='color:#3A6D8C'>Struktur Data Pandas</span></b>

* Terdapat dua struktur data utama pada library pandas yang disebut dengan Series dan DataFrame.
<p align="center">
<img src="https://datagy.io/wp-content/uploads/2022/12/Understanding-Pandas-Series-and-DataFrames-1024x576.png" width="550" height="289">
<p align="center"><b>Gambar 2. Series VS DataFrame</b></p>
<br>

* Gabungan beberapa Series yang mempunyai banyak baris yang sama dapat membentuk sebuah DataFrame atau berlaku sebaliknya pada sebuah DataFrame dapat dipecah (di-<i>slicing</i> satu kolom) membentuk sebuah Series.<br><br>

<ul><p align="left"><sup>docs : <i><a href="https://shorturl.at/dezQY">http://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html</i></sup></a></p></ul>

## <b>B.1. <span style='color:#6A9AB0'>Series</span></b>  

* <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html">Series</a> merupakan kolom dari sebuah tabel yang bersifat homogen atau mempunyai elemen dengan tipe data yang sejenis.

* Untuk membuat Series gunakan sintaks berikut

<p align="center"><code><b>pd.Series(data)</b></code></p>

<ul>
  data dapat berupa <i>array-like</i>, <i>Iterable</i>, <i>dict</i>, atau <i>scalar</i>
</ul>
<br>
<ul><p align="left"><sup>docs : <i><a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html">https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html</i></sup></a></p></ul>



###<b><span style='color:#55679C'>Quest B.1.1</span></b>
<i>Hanya Singapura dan Indonesia yang saat ini telah menetapkan pajak karbon (carbon pricing) di Asia Tenggara. (source : https://accept.aseanenergy.org/progress-of-carbon-pricing-in-asean)

Buatlah sebuah Series yang disimpan pada variabel dengan nama `country_carbon_pricing` untuk menampung nama negara yang telah menetapkan pajak karbon di Asia Tenggara, lalu tampilkan outputnya!</i>

In [ ]:
# Answer B.1.1
import pandas as pd
countries = ["Singapore", "Indonesia"]
country_carbon_pricing = pd.Series(countries)
print(country_carbon_pricing)

0    Singapore
1    Indonesia
dtype: object


###<b><span style='color:#55679C'>Quest B.1.2</span></b>
<i>Dari data emisi karbon yang tercatat di Indonesia, akan diberi detail data sebagai berikut

* Tahun 2015 dihasilkan 491.46 Mton CO2
* Tahun 2016 dihasilkan 488.40 Mton CO2
* Tahun 2017 dihasilkan 511.33 Mton CO2
* Tahun 2018 dihasilkan 579.23 Mton CO2
* Tahun 2019 dihasilkan 625.66 Mton CO2

<i>Source: Joint Research Centre of European Union</i>

Dari data tersebut buatlah sebuah Series yang disimpan pada variabel dengan nama `emisi_co2_indonesia` lalu tampilkan outputnya!</i>

In [ ]:
# Answer B.1.2
import pandas as pd
data_emisi = {"Tahun 2015": 491.46,
    "Tahun 2016": 488.40,
    "Tahun 2017": 511.33,
    "Tahun 2018": 579.23,
    "Tahun 2019": 625.66

}
emisi_co2_indonesia = pd.Series(data_emisi)
print(emisi_co2_indonesia)

Tahun 2015    491.46
Tahun 2016    488.40
Tahun 2017    511.33
Tahun 2018    579.23
Tahun 2019    625.66
dtype: float64


## <b>B.2. <span style='color:#6A9AB0'>DataFrame</span></b>  

* <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html#pandas.DataFrame">DataFrame</a> adalah struktur data 2 dimensi yang terdiri dari baris dan kolom (seperti tabel) dan dapat menyimpan data dari berbagai jenis (termasuk karakter, bilangan bulat, bilangan real, data kategorikal, dan lainnya).

<p align="center">
<img src="https://www.w3resource.com/w3r_images/pandas-data-frame.svg" width="600" height="300">
<p align="center"><b>Gambar 2. DataFrame Pandas</b></p>
<br>

* <a href="https://pandas.pydata.org/docs/reference/api/pandas.Series.html#pandas.Series">Series</a> merupakan kolom dari sebuah tabel yang bersifat homogen atau mempunyai elemen dengan tipe data yang sejenis.


<p align="center">
<img src="https://datagy.io/wp-content/uploads/2022/12/Understanding-Pandas-Series-and-DataFrames-1024x576.png" width="550" height="300">
<p align="center"><b>Gambar 3. Series VS DataFrame</b></p>
<br>

* Gabungan beberapa Series yang mempunyai banyak baris yang sama dapat membentuk sebuah DataFrame atau berlaku sebaliknya pada sebuah DataFrame dapat dipecah (di-<i>slicing</i> satu kolom) membentuk sebuah Series.<br><br>

* Untuk membuat sebuah DataFrame secara manual, dapat menggunakan sintaks
<p align="center"><code><b>pd.DataFrame(data)</b></code></p>
<ul>data inputan dapat berupa ndarray (structured atau homogeneous), iterable atau dict</ul>
<br>
<ul><p align="left"><sup>docs : <i><a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html">https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html</i></sup></a></p></ul>



###<b><span style='color:#55679C'>Quest B.2</span></b>
<i>Diberikan data Air Quality Index (AQI) dan partikel udara yang memiliki diameter kurang dari 2,5 mikrometer (PM2.5) di Jakarta (source : <a href="https://www.iqair.com/id/indonesia/jakarta?srsltid=AfmBOoqcwssGzLEzHMeGJwjaU-oi7wh1JeUqR-wBxicstXuvUN_i5UwT">https://www.iqair.com/id/indonesia/jakarta?srsltid=AfmBOoqcwssGzLEzHMeGJwjaU-oi7wh1JeUqR-wBxicstXuvUN_i5UwT</a>) sebagai berikut :</i>
<br><br>
<table>
    <tr>
        <th>Tanggal</th>
        <th>PM2.5</th>
        <th>AQI</th>
    </tr>
    <tr>
        <td>2024-09-21</td>
        <td>70.7</td>
        <td>162</td>
    </tr>
    <tr>
        <td>2024-09-22</td>
        <td>58.6</td>
        <td>153</td>
    </tr>
    <tr>
        <td>2024-09-23</td>
        <td>52.3</td>
        <td>142</td>
    </tr>
</table>
<br>

<i>Dari data diatas, buatkan secara manual DataFrame-nya dengan pandas!</i>



In [ ]:
# Answer B.2
import pandas as pd


data = pd.DataFrame({
    'tanggal': ['2024-09-21', '2024-09-22', '2024-09-23'],
    'pm2.5': [70.7, 58.6, 52.3],
    'aqi': [162, 153, 142]
})
display(data)

tanggal  pm2.5  aqi
0  2024-09-21   70.7  162
1  2024-09-22   58.6  153
2  2024-09-23   52.3  142

## <b>B.3. <span style='color:#6A9AB0'>Parts Of DataFrame</span></b>

Untuk lebih memahami objek DataFrame, perlu diketahui bahwa DataFrame terdiri dari tiga komponen sebagai atribut berikut :

<ul>1️⃣ .values : Nilai - nilai yang tertera pada DataFrame</ul>
<ul>2️⃣ .columns : Nama kolom yang ada pada DataFrame</ul>
<ul>3️⃣ .index : Nama atau nomor index pada DataFrame</ul>



###<b><span style='color:#55679C'>Quest B.3.1</span></b>

<i>Pada DataFrame yang telah dibuat pada Quest B.2, coba buatkan list nama kolomnya!</i>


In [ ]:
# Answer B.3.1
nama_kolom = data.columns

print(nama_kolom)

Index(['tanggal', 'pm2.5', 'aqi'], dtype='object')


###<b><span style='color:#55679C'>Quest B.3.2</span></b>

<i>Pada DataFrame yang telah dibuat pada Quest B.2, coba akses setiap nilainya!</i>


In [ ]:
# Answer B.3.2
semua_nilai = data.values
print("Semua nilai dalam DataFrame:")
print(semua_nilai)

Semua nilai dalam DataFrame:
[['2024-09-21' 70.7 162]
 ['2024-09-22' 58.6 153]
 ['2024-09-23' 52.3 142]]


# <b>C. <span style='color:#3A6D8C'>Data Manipulation</span></b>

Data Manipulation adalah sebuah metode pengorganisasian data agar lebih mudah dibaca, terstruktur dan agar lebih mudah diolah untuk kebutuhan lain kedepannya. Beberapa hal yang biasa dilakukan saat data manipulation adalah sebagai berikut :


<ul>1️⃣ Column selections</ul>
<ul>2️⃣ Filter the data</ul>
<ul>3️⃣ Create a new column or change it</ul>
<ul>4️⃣ Drop a column</ul>
<ul>5️⃣ Apply custom functions</ul>
<ul>6️⃣ Join / Merge the data</ul>
<ul>7️⃣ etc</ul>


## <b>C.1. <span style='color:#6A9AB0'>Data Extraction</span></b>

* Dalam konteks data extraction, Pandas dapat digunakan untuk membaca data dari berbagai sumber, seperti file CSV, Excel, SQL database, dan lainnya, dan mengubahnya menjadi DataFrames untuk analisis lebih lanjut.

* Beberapa fungsi dan metode Pandas yang umum digunakan untuk data extraction adalah sebagai berikut:

<ul>
<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html">pd.read_csv()</a></code>: Digunakan untuk membaca data dari file CSV dan mengonversinya menjadi DataFrame.

<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html">pd.read_excel()</a></code>: Membaca data dari file Excel dan membuat DataFrame dari lembar kerja atau tabel tertentu.

<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html">pd.read_sql()</a></code>: Mengambil data dari basis data SQL menggunakan SQL query dan menyimpannya dalam bentuk DataFrame.
</ul>


###<b><span style='color:#55679C'>Quest C.1.</span></b>

<i>Untuk melakukan proses ekstraksi data, lakukan instruksi berikut :     

1.   Persiapkan dataset WorldSustainabilityDataset.csv
2.   Upload ke google colab dengan cara pilih File pada bagian kiri laman (icon 📁), kemudian klik kanan dan pilih upload
3.   Lakukan ekstraksi data csv (.csv) dengan library pandas, assign hasilnya ke sebuah variabel bernama <code>data_world_sust</code>
4.   Tampilkan hasilnya</i>






In [ ]:
# Answer C.1.
# 2. assign hasilnya ke sebuah variabel bernama data_world_sust
data_world_sust = pd.read_csv('/content/WorldSustainabilityDataset.csv')

# 3. Menampilkan Hasil
display(data_world_sust)

Country Name Country Code  Year  \
0           Aruba          ABW  2000   
1           Aruba          ABW  2001   
2           Aruba          ABW  2002   
3           Aruba          ABW  2003   
4           Aruba          ABW  2004   
...           ...          ...   ...   
3282     Zimbabwe          ZWE  2014   
3283     Zimbabwe          ZWE  2015   
3284     Zimbabwe          ZWE  2016   
3285     Zimbabwe          ZWE  2017   
3286     Zimbabwe          ZWE  2018   

     Access to electricity (% of population) - EG.ELC.ACCS.ZS  \
0                                            91.660.398         
1                                                   100         
2                                                   100         
3                                                   100         
4                                                   100         
...                                                 ...         
3282                                               32.3         
3283                                               33.7         
3284                                      3.967.622.757         
3285                                      4.014.428.329         
3286                                      4.061.635.971         

     Adjusted net national income per capita (annual % growth) - NY.ADJ.NNTY.PC.KD.ZG  \
0                                                   NaN                                 
1                                                   NaN                                 
2                                                   NaN                                 
3                                                   NaN                                 
4                                                   NaN                                 
...                                                 ...                                 
3282                                        763.955.299                                 
3283                                      1.074.900.886                                 
3284                                        -71.604.768                                 
3285                                      2.817.257.174                                 
3286                                      6.155.657.985                                 

     Adjusted net savings, excluding particulate emission damage (% of GNI) - NY.ADJ.SVNX.GN.ZS  \
0                                         1.539.221.136                                           
1                                         1.149.193.045                                           
2                                         3.893.619.865                                           
3                                           671.386.335                                           
4                                         9.801.575.575                                           
...                                                 ...                                           
3282                                     -1.409.402.347                                           
3283                                     -1.888.283.086                                           
3284                                     -1.119.921.264                                           
3285                                     -1.270.023.821                                           
3286                                                NaN                                           

     Adjusted savings: carbon dioxide damage (% of GNI) - NY.ADJ.DCO2.GN.ZS  \
0                                         2.389.134.359                       
1                                         2.509.045.841                       
2                                           272.138.537                       
3                                         2.733.675.941                       
4                                         2.678.363.121                       
...        

## <b>C.2. <span style='color:#6A9AB0'>Inspecting Data Information</span></b>

* Untuk dapat menginspeksi informasi umum pada data dapat menggunakan sintaks berikut

<p align="center"><code><b>DataFrame.info()</b></code></p>

<ul>Ketika dijalankan, sintaks tersebut akan menghasilkan output informasi umum pada data seperti banyak kolom, nama kolom beserta tipe data dan jumlah data tak kosong (<i>non-null value</i>)</ul><br>

<ul><p align="left"><sup>docs : <i><a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html</i></sup></a></p></ul>


### <b><span style='color:#55679C'>Quest C.2</span></b>

<i>Inspeksi informasi umum pada data yang telah diekstraksi dan telah disimpan pada variabel <code>data_world_sust</code>. Setelah itu coba uraikan informasi apa yang bisa didapatkan setelah melakukan inspeksi informasi umum pada <code>data_world_sust</code>?</i>

In [ ]:
# Answer C.2
data_world_sust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3287 entries, 0 to 3286
Data columns (total 54 columns):
 #   Column                                                                                        Non-Null Count  Dtype  
---  ------                                                                                        --------------  -----  
 0   Country Name                                                                                  3287 non-null   object 
 1   Country Code                                                                                  3287 non-null   object 
 2   Year                                                                                          3287 non-null   int64  
 3   Access to electricity (% of population) - EG.ELC.ACCS.ZS                                      3193 non-null   object 
 4   Adjusted net national income per capita (annual % growth) - NY.ADJ.NNTY.PC.KD.ZG              2566 non-null   object 
 5   Adjusted net savings, exclu

## <b>C.3. <span style='color:#6A9AB0'>Data Selection</span></b>

* Untuk memilih kolom tertentu pada sebuah DataFrame dapat menggunakan sintaks berikut

<p align="center"><code><b>DataFrame.loc[baris : nama_kolom]</b></code></p>

* <code>baris</code>
<ul>
  <li>Gunakan simbol ( : ) untuk menampilkan semua data</li>
  <li>Gunakan simbol (m : n) untuk menampilkan data dari index ke - m sampai dengan index ke - n</li>
</ul>

* <code>nama_kolom</code>
<ul>
  <li>Gunakan string untuk mengakses satu kolom saja (<i>return : Series</i>)</li>
  <li>Gunakan list untuk mengakses lebih dari satu kolom  (<i>return : DataFrame</i>)</li>
</ul>


###<b><span style='color:#55679C'>Quest C.3.1</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, simpan 18 data teratas ke dalam variabel **year** untuk kolom **Year** saja. Kemudian tampilkan data tersebut!</i>




In [ ]:
# Answer C.3.1
year = data_world_sust.loc[0:18, ['Year']]
display(year)
print(f"\ndtype: {year['Year'].dtype}")

Year
0   2000
1   2001
2   2002
3   2003
4   2004
5   2005
6   2006
7   2007
8   2008
9   2009
10  2010
11  2011
12  2012
13  2013
14  2014
15  2015
16  2016
17  2017
18  2018


dtype: int64


###<b><span style='color:#55679C'>Quest C.3.2</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, Tampilkan 18 data teratas untuk kolom
* Country Name
* Year
* Access to electricity (% of population) - EG.ELC.ACCS.ZS</i>

In [ ]:
# Answer C.3.2


Country Name  Year  \
0         Aruba  2000   
1         Aruba  2001   
2         Aruba  2002   
3         Aruba  2003   
4         Aruba  2004   
5         Aruba  2005   
6         Aruba  2006   
7         Aruba  2007   
8         Aruba  2008   
9         Aruba  2009   
10        Aruba  2010   
11        Aruba  2011   
12        Aruba  2012   
13        Aruba  2013   
14        Aruba  2014   
15        Aruba  2015   
16        Aruba  2016   
17        Aruba  2017   
18        Aruba  2018   

    Access to electricity (% of population) - EG.ELC.ACCS.ZS  
0                                           91.660398         
1                                          100.000000         
2                                          100.000000         
3                                          100.000000         
4                                          100.000000         
5                                          100.000000         
6                                          100.000000         
7                                          100.000000         
8                                          100.000000         
9                                          100.000000         
10                                          93.356292         
11                                         100.000000         
12                                         100.000000         
13                                         100.000000         
14                                         100.000000         
15                                         100.000000         
16                                         100.000000         
17                                         100.000000         
18                                         100.000000

## <b>C.4. <span style='color:#6A9AB0'>Conditions</span></b>

* Salah satu teknik paling sederhana untuk menemukan bagian mana dari kumpulan data yang menarik adalah menemukan subset baris yang cocok dengan beberapa kriteria untuk difilter atau dipilih. Untuk melakukan hal ini cukup mudah yakni dengan menuliskan kondisi seperti berikut

<p align="center"><code><b>DataFrame.loc[conditions]</b></code></p>

* Conditions dapat dibentuk dari 1 atau lebih kriteria boolean dan dihubungkan dengan operator logika and (&), or (|) atau not (~)

<br>

docs : <i>https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html</i>

###<b><span style='color:#55679C'>Quest C.4.1</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, tampilkan semua data yang Country Name-nya Indonesia</i>

In [ ]:
# Answer C.4.1
data_Indonesia = data_world_sust.loc[data_world_sust['Country Name'] == 'Indonesia']
display(data_Indonesia)

Country Name Country Code  Year  \
1368    Indonesia          IDN  2000   
1369    Indonesia          IDN  2001   
1370    Indonesia          IDN  2002   
1371    Indonesia          IDN  2003   
1372    Indonesia          IDN  2004   
1373    Indonesia          IDN  2005   
1374    Indonesia          IDN  2006   
1375    Indonesia          IDN  2007   
1376    Indonesia          IDN  2008   
1377    Indonesia          IDN  2009   
1378    Indonesia          IDN  2010   
1379    Indonesia          IDN  2011   
1380    Indonesia          IDN  2012   
1381    Indonesia          IDN  2013   
1382    Indonesia          IDN  2014   
1383    Indonesia          IDN  2015   
1384    Indonesia          IDN  2016   
1385    Indonesia          IDN  2017   
1386    Indonesia          IDN  2018   

     Access to electricity (% of population) - EG.ELC.ACCS.ZS  \
1368                                               86.3         
1369                                              86.26         
1370                                               87.6         
1371                                              87.94         
1372                                              89.01         
1373                                      8.542.111.206         
1374                                              90.62         
1375                                               91.1         
1376                                              92.73         
1377                                              93.55         
1378                                              94.15         
1379                                              94.83         
1380                                                 96         
1381                                      9.646.425.831         
1382                                              97.01         
1383                                      9.753.736.744         
1384                                              97.62         
1385                                              98.14         
1386                                              98.51         

     Adjusted net national income per capita (annual % growth) - NY.ADJ.NNTY.PC.KD.ZG  \
1368                                      1.750.944.808                                 
1369                                      2.974.086.337                                 
1370                                     -1.288.692.747                                 
1371                                     -4.460.944.746                                 
1372                                     -8.841.897.946                                 
1373                                      2.294.930.782                                 
1374                                     -5.243.976.158                                 
1375                                        -64.565.529                                 
1376                                     -2.618.162.435                                 
1377                                      2.201.790.228                                 
1378                                      1.319.570.875                                 
1379                                        422.890.557                                 
1380                                      3.675.769.651                                 
1381                                      4.185.755.853                                 
1382                                      2.529.367.368                                 
1383                                      5.124.024.959                                 
1384                                      6.383.631.021                                 
1385                                      4.928.757.596                                 
1386                                      4.733.578.245                                 

     Adjusted net savings, excluding particulate emission damage (% of GNI) - NY.ADJ.SVNX.GN.ZS  \
1368                   

###<b><span style='color:#55679C'>Quest C.4.2</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, tampilkan semua data yang Country Name-nya Indonesia namun hanya untuk tahun 2012 keatas saja</i>

In [ ]:
# Answer C.4.2
data_Indonesia = data_world_sust.loc[(data_world_sust['Country Name'] == 'Indonesia') & (data_world_sust['Year'] >= 2012)]
display(data_Indonesia)

Country Name Country Code  Year  \
1380    Indonesia          IDN  2012   
1381    Indonesia          IDN  2013   
1382    Indonesia          IDN  2014   
1383    Indonesia          IDN  2015   
1384    Indonesia          IDN  2016   
1385    Indonesia          IDN  2017   
1386    Indonesia          IDN  2018   

     Access to electricity (% of population) - EG.ELC.ACCS.ZS  \
1380                                                 96         
1381                                      9.646.425.831         
1382                                              97.01         
1383                                      9.753.736.744         
1384                                              97.62         
1385                                              98.14         
1386                                              98.51         

     Adjusted net national income per capita (annual % growth) - NY.ADJ.NNTY.PC.KD.ZG  \
1380                                      3.675.769.651                                 
1381                                      4.185.755.853                                 
1382                                      2.529.367.368                                 
1383                                      5.124.024.959                                 
1384                                      6.383.631.021                                 
1385                                      4.928.757.596                                 
1386                                      4.733.578.245                                 

     Adjusted net savings, excluding particulate emission damage (% of GNI) - NY.ADJ.SVNX.GN.ZS  \
1380                                      1.132.974.606                                           
1381                                      1.123.264.243                                           
1382                                      1.110.148.128                                           
1383                                      1.179.867.101                                           
1384                                        136.911.197                                           
1385                                      1.472.010.401                                           
1386                                      1.533.350.602                                           

     Adjusted savings: carbon dioxide damage (% of GNI) - NY.ADJ.DCO2.GN.ZS  \
1380                                      1.844.120.111                       
1381                                      1.514.860.592                       
1382                                      1.445.982.622                       
1383                                      2.141.862.752                       
1384                                      2.243.017.853                       
1385                                      2.181.155.105                       
1386                                      2.273.700.403                       

     Adjusted savings: natural resources depletion (% of GNI) - NY.ADJ.DRES.GN.ZS  \
1380                                      3.709.363.395                             
1381                                      3.308.080.828                             
1382                                      2.773.960.097                             
1383                                      1.650.661.598                             
1384                                      1.383.614.014                             
1385                                      1.686.334.756                             
1386                                      2.521.512.898                             

     Adjusted savings: net forest depletion (% of GNI) - NY.ADJ.DFOR.GN.ZS  \
1380                                                  0                      
1381                                                  0                      
1382                                                  0                      
1383                           

###<b><span style='color:#55679C'>Quest C.4.3</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, tampilkan kolom 'Country Name', 'GDP per capita (current US$) - NY.GDP.PCAP.CD', 'Inflation, consumer prices (annual %) - FP.CPI.TOTL.ZG' yang Country Name-nya Indonesia</i>

In [ ]:
# Answer C.4.3
data_filter = data_world_sust.loc[data_world_sust['Country Name'] == 'Indonesia', ['Country Name', 'GDP per capita (current US$) - NY.GDP.PCAP.CD', 'Inflation, consumer prices (annual %) - FP.CPI.TOTL.ZG']]
display(data_filter)

Country Name GDP per capita (current US$) - NY.GDP.PCAP.CD  \
1368    Indonesia                                 7.801.902.047   
1369    Indonesia                                 7.482.576.087   
1370    Indonesia                                 9.001.775.881   
1371    Indonesia                                   106.564.852   
1372    Indonesia                                 1.150.261.367   
1373    Indonesia                                 1.263.287.332   
1374    Indonesia                                 1.589.801.489   
1375    Indonesia                                 1.860.002.811   
1376    Indonesia                                 2.166.854.231   
1377    Indonesia                                 2.261.247.309   
1378    Indonesia                                 3.122.362.673   
1379    Indonesia                                 3.643.047.176   
1380    Indonesia                                   369.435.934   
1381    Indonesia                                 3.623.927.241   
1382    Indonesia                                 3.491.637.491   
1383    Indonesia                                 3.331.695.115   
1384    Indonesia                                 3.562.816.334   
1385    Indonesia                                 3.837.578.024   
1386    Indonesia                                 3.893.859.578   

     Inflation, consumer prices (annual %) - FP.CPI.TOTL.ZG  
1368                                        368.861.916      
1369                                      1.150.011.488      
1370                                      1.190.011.757      
1371                                      6.757.317.194      
1372                                      6.064.059.885      
1373                                      1.045.319.842      
1374                                        131.086.721      
1375                                      6.406.562.813      
1376                                      1.022.666.455      
1377                                        438.641.555      
1378                                      5.134.204.008      
1379                                        535.604.779      
1380                                      4.279.499.996      
1381                                      6.412.513.302      
1382                                      6.394.925.408      
1383                                      6.363.121.131      
1384                                      3.525.805.157      
1385                                        380.879.807      
1386                                      3.198.346.416

###<b><span style='color:#55679C'>Quest C.4.4</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, Pada tahun 2018, bandingkan GDP per capita pada negara Indonesia, Singapura, Malaysia dan Brunei Darussalam. Tampilkan kolom 'Country Name', 'Year', 'GDP per capita (current US$) - NY.GDP.PCAP.CD' saja </i>

In [ ]:
# Answer C.4.4
filter_data = data_world_sust.loc[
    (data_world_sust['Country Name'].isin(['Indonesia', 'Singapore', 'Malaysia', 'Brunei Darussalam'])) &
    (data_world_sust['Year'] == 2018),
    ['Country Name', 'Year', 'GDP per capita (current US$) - NY.GDP.PCAP.CD']
]

display(filter_data)

Country Name  Year GDP per capita (current US$) - NY.GDP.PCAP.CD
474   Brunei Darussalam  2018                                 3.162.847.626
1386          Indonesia  2018                                 3.893.859.578
2184           Malaysia  2018                                   113.776.542
2659          Singapore  2018                                 6.667.904.649

## <b>C.5. <span style='color:#6A9AB0'>Sorting Row</span></b>

Menemukan data yang menarik dalam sebuah DataFrame seringkali lebih mudah jika dilakukan pengurutan baris datanya. Pengurutan dibagi menjadi dua yakni pengurutan secara `Ascending` (pengurutan naik A → Z) dan juga pengurutan secara `Descending` (pengurutan turun Z → A). Eksekusi sintaks berikut :

<p align="center"><code><b>DataFrame.sort_values(by, ascending=True, ignore_index=False)</b></code></p>

Jika ingin mengurutkan DataFrame pandas dengan memperhatikan argumen

<ul>
<li><b><i>by</i></b> sebagai satu atau lebih kolom yang ingin dijadikan patokan pengurutan</li>

<li><b><i>ascending</i></b> dapat diisikan True jika ingin terurut naik atau False jika ingin terurut turun (dafault : True)</li>

<li><b><i>ignore_index</i></b> dapat diisikan True jika ingin index masih terurut atau False jika indexnya masih dipertahankan sesuai data awal (dafault : False)</li>

</ul><br>

docs : <i>https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html</i>


###<b><span style='color:#55679C'>Quest C.5.1</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, tampilkan semua data yang Country Name-nya Indonesia namun urutkan dari GDP dari yang terbesar sampai yang terkecil</i>

In [ ]:
# Answer C.5.1
pd.set_option('display.max_columns', None)
data_indonesia = data_world_sust.loc[data_world_sust['Country Name'] == 'Indonesia']
# Mengurutkan berdasarkan GDP desc
indonesia_sorted = data_indonesia.sort_values(by='GDP (current US$) - NY.GDP.MKTP.CD', ascending=False)
display(indonesia_sorted)

Country Name Country Code  Year  \
1386    Indonesia          IDN  2018   
1385    Indonesia          IDN  2017   
1384    Indonesia          IDN  2016   
1380    Indonesia          IDN  2012   
1381    Indonesia          IDN  2013   
1379    Indonesia          IDN  2011   
1382    Indonesia          IDN  2014   
1383    Indonesia          IDN  2015   
1378    Indonesia          IDN  2010   
1377    Indonesia          IDN  2009   
1376    Indonesia          IDN  2008   
1375    Indonesia          IDN  2007   
1374    Indonesia          IDN  2006   
1373    Indonesia          IDN  2005   
1372    Indonesia          IDN  2004   
1371    Indonesia          IDN  2003   
1370    Indonesia          IDN  2002   
1368    Indonesia          IDN  2000   
1369    Indonesia          IDN  2001   

     Access to electricity (% of population) - EG.ELC.ACCS.ZS  \
1386                                              98.51         
1385                                              98.14         
1384                                              97.62         
1380                                                 96         
1381                                      9.646.425.831         
1379                                              94.83         
1382                                              97.01         
1383                                      9.753.736.744         
1378                                              94.15         
1377                                              93.55         
1376                                              92.73         
1375                                               91.1         
1374                                              90.62         
1373                                      8.542.111.206         
1372                                              89.01         
1371                                              87.94         
1370                                               87.6         
1368                                               86.3         
1369                                              86.26         

     Adjusted net national income per capita (annual % growth) - NY.ADJ.NNTY.PC.KD.ZG  \
1386                                      4.733.578.245                                 
1385                                      4.928.757.596                                 
1384                                      6.383.631.021                                 
1380                                      3.675.769.651                                 
1381                                      4.185.755.853                                 
1379                                        422.890.557                                 
1382                                      2.529.367.368                                 
1383                                      5.124.024.959                                 
1378                                      1.319.570.875                                 
1377                                      2.201.790.228                                 
1376                                     -2.618.162.435                                 
1375                                        -64.565.529                                 
1374                                     -5.243.976.158                                 
1373                                      2.294.930.782                                 
1372                                     -8.841.897.946                                 
1371                                     -4.460.944.746                                 
1370                                     -1.288.692.747                                 
1368                                      1.750.944.808                                 
1369                                      2.974.086.337                                 

     Adjusted net savings, excluding particulate emission damage (% of GNI) - NY.ADJ.SVNX.GN.ZS  \
1386                   

###<b><span style='color:#55679C'>Quest C.5.2</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, pada tahun 2018, negara mana saja yang GDPnya masuk kedalam 10 terbaik? Tampilkan hanya kolom 'Country Name', 'Year', 'GDP per capita (current US$) - NY.GDP.PCAP.CD'  </i>

In [ ]:
# Answer C.5.2
top_10_gdp = (data_world_sust[data_world_sust['Year'] == 2018]
              [['Country Name', 'Year', 'GDP per capita (current US$) - NY.GDP.PCAP.CD']]
              .sort_values(by='GDP per capita (current US$) - NY.GDP.PCAP.CD', ascending=False)
              .head(10)
              .reset_index(drop=True))
display(top_10_gdp)

Country Name  Year GDP per capita (current US$) - NY.GDP.PCAP.CD
0     Bulgaria  2018                                   942.773.043
1        China  2018                                 9.976.677.137
2   Kazakhstan  2018                                 9.812.626.371
3       Mexico  2018                                 9.686.514.244
4       Turkey  2018                                 9.453.196.172
5       Brazil  2018                                 9.151.445.253
6         Togo  2018                                 9.015.230.169
7         Mali  2018                                 8.948.047.765
8   Montenegro  2018                                 8.850.094.631
9         Cuba  2018                                 8.824.194.008

###<b><span style='color:#55679C'>Quest C.5.3</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, pada tahun 2016, negara mana saja yang Emisi karbon dioksida (CO2) berbasis produksi tahunannya masuk dalam 5 negara penghasil emisi CO2 terkecil? Tampilkan nama negara, tahun dan jumlah emisi karbonnya! </i>

In [ ]:
# Answer C.5.3
smallest_emitters_2016_top5 = (
    data_world_sust.loc[data_world_sust['Year'] == 2016]
      [['Country Name', 'Year', 'Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes']]
      .sort_values(by='Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes',
                   ascending=True
                   )
      .head(5)
)
display(smallest_emitters_2016_top5) # Changed from smallest_emitters_2016_top10

Country Name  Year  \
852      Dominica  2016   
966       Eritrea  2016   
1840   Luxembourg  2016   
2600       Rwanda  2016   
491        Bhutan  2016   

     Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes  
852                                                0.18                                     
966                                                0.66                                     
1840                                              09.08                                     
2600                                              1.047                                     
491                                               1.153

###<b><span style='color:#55679C'>Quest C.5.4</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, pada tahun 2018 di benua Asia, tampilkan 10 negara dengan produksi Emisi CO2 terbesar. Urutkan dari yang terbesar hingga terkecil</i>

In [ ]:
# Answer C.5.4
df= data_world_sust[
    (data_world_sust['Year'] == 2018) &
    (data_world_sust['Continent'] == 'Asia')
]

df = df.sort_values(
    by='Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes',
    ascending=False,
)


df_final = df[['Country Name', 'Year', 'Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes']]


df_final.head(10)

Country Name  Year  \
607                China  2018   
1139             Georgia  2018   
474    Brunei Darussalam  2018   
284           Bangladesh  2018   
1443  Iran, Islamic Rep.  2018   
2355                Oman  2018   
2944          Tajikistan  2018   
2089            Mongolia  2018   
1671         Korea, Rep.  2018   
1500              Israel  2018   

     Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes  
607                                           9.956.569                                     
1139                                              9.862                                     
474                                                9.56                                     
284                                              85.719                                     
1443                                            755.402                                     
2355                                              71.03                                     
2944                                              7.473                                     
2089                                             64.508                                     
1671                                            634.934                                     
1500                                             62.213

## <b>C.6. <span style='color:#6A9AB0'>Summarize Data</span></b>

Di pandas, method `DataFrame.describe()` digunakan untuk menghasilkan statistik deskriptif dari DataFrame. Metode ini sangat berguna untuk dengan cepat memahami statistik dasar dari data. Berikut adalah statistik utama yang disediakan :

* `Jumlah` : Jumlah nilai non-null dalam setiap kolom.
* `Rata-rata` : Rata-rata aritmetika dari setiap kolom.
* `Standar Deviasi` : Ukuran dispersi nilai-nilai sekitar rata-rata.
* `Minimum` : Nilai minimum dalam setiap kolom.
* `Kuartil 25 (Q1)` : Nilai di bawahnya 25% data berada.
* `Kuartil 50 (Median atau Q2)` : Nilai di bawahnya 50% data berada (yaitu nilai tengah).
* `Kuartil 75 (Q3)` : Nilai di bawahnya 75% data berada.
* `Maksimum` : Nilai maksimum dalam setiap kolom.

###<b><span style='color:#55679C'>Quest C.6</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, hitung semua statistik deskriptif menggunakan .describe() pada kolom numerik!</i>

In [ ]:
# Answer C.6
df= data_world_sust[
    (data_world_sust['Year'] == 2018) &
    (data_world_sust['Continent'] == 'Asia')
]

df = df.sort_values(
    by='Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes',
    ascending=False,
)


df_final = df[['Country Name', 'Year', 'Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes']]


df_final.head(10)

Country Name  Year  \
607                China  2018   
1139             Georgia  2018   
474    Brunei Darussalam  2018   
284           Bangladesh  2018   
1443  Iran, Islamic Rep.  2018   
2355                Oman  2018   
2944          Tajikistan  2018   
2089            Mongolia  2018   
1671         Korea, Rep.  2018   
1500              Israel  2018   

     Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes  
607                                           9.956.569                                     
1139                                              9.862                                     
474                                                9.56                                     
284                                              85.719                                     
1443                                            755.402                                     
2355                                              71.03                                     
2944                                              7.473                                     
2089                                             64.508                                     
1671                                            634.934                                     
1500                                             62.213

## <b>C.7. <span style='color:#6A9AB0'>Aggregation</span></b>

Agregasi data merupakan serangkaian proses yang dilakukan untuk melakukan pengumpulan data dan dinyatakan dalam bentuk ringkasan sebagai contoh menghitung total yang dikelompokkan berdasarkan kategori tertentu, dsb.

Untuk melakukan agregasi gunakan sintaks berikut

<p align="center"><code><b>DataFrame.group_by(by, as_index).agg(kolom_hasil_aggregasi = (kolom_dihitung, jenis_agregasi))</b></code></p>

Jenis agregasi :
* sum = Menghitung total data
* mean = Menghitung rataan data
* max = Menghitung nilai maksimum data
* min = Menghitung nilai minimum data
* nunique = Menghitung jumlah data unik data
* count = Menghitung banyak data

###<b><span style='color:#55679C'>Quest C.7.1</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, hitung rata-rata GDP tiap Benua pada seluruh tahun!</i>

In [ ]:
# Answer C.4.1
ratarata_gdp_per_benua = data_world_sust.groupby('Continent')['GDP (current US$) - NY.GDP.MKTP.CD'].mean()
display(ratarata_gdp_per_benua)

Continent
Africa           3.335216e+10
Asia             4.218820e+11
Europe           4.504028e+11
North America    8.040119e+11
Oceania          1.907884e+11
South America    2.405237e+11
Name: GDP (current US$) - NY.GDP.MKTP.CD, dtype: float64

###<b><span style='color:#55679C'>Quest C.7.2</span></b>

<i>Pada DataFrame <code>data_world_sust</code>, dengan cara agregasi buatkan 2 insight lainnya yang menurut kamu perlu dianalisa sertakan juga insight yang kamu peroleh!</i>

In [2]:
import pandas as pd

# Membaca data dari GitHub
url = 'https://raw.githubusercontent.com/PineTrader/Data/refs/heads/main/WorldSustainabilityDataset.csv'
data_world_sust = pd.read_csv(url)
# Menghitung rata-rata Pengeluaran Pemerintah untuk Konsumsi Akhir per Benua
average_gov_exp_per_continent = data_world_sust.groupby('Continent', as_index=False).agg(
    rata_rata_pengeluaran_pemerintah=('General government final consumption expenditure (% of GDP) - NE.CON.GOVT.ZS', 'mean')
)

# Menampilkan hasil
display(average_gov_exp_per_continent)

Continent  rata_rata_pengeluaran_pemerintah
0         Africa                         15.018418
1           Asia                         15.421302
2         Europe                         19.287896
3  North America                         15.308493
4        Oceania                         18.516184
5  South America                         13.799687

In [6]:
import pandas as pd

# Membaca data dari GitHub
url = 'https://raw.githubusercontent.com/PineTrader/Data/refs/heads/main/WorldSustainabilityDataset.csv'
data_world_sust = pd.read_csv(url)

# Menghitung rata-rata Penggunaan Energi Terbarukan per Benua
average_renewable_energy_per_continent = data_world_sust.groupby('Continent', as_index=False).agg(
    rata_rata_penggunaan_energi_terbarukan=('Renewable energy consumption (% of total final energy consumption) - EG.FEC.RNEW.ZS', 'mean')
)

# Menampilkan hasil
display("Rata-rata Penggunaan Energi Terbarukan per Benua:")
for index, row in average_renewable_energy_per_continent.iterrows():
    display(f"{row['Continent']}: {row['rata_rata_penggunaan_energi_terbarukan']:.2f}%")

'Rata-rata Penggunaan Energi Terbarukan per Benua:'

'Africa: 61.45%'

'Asia: 21.31%'

'Europe: 18.77%'

'North America: 23.07%'

'Oceania: 28.27%'

'South America: 29.87%'

Insight :

Rata-rata pengeluaran pemeritah paling banyak dibenua Europe dengan rata-rata pengeluaran 19.287896. Sedangkan untuk rata-rata pengeluaran pemerintah paling sedikt dibenua South America	dengan rata-rata pengeluaran 13.799687.

rata-rata penggunaan energi terbarukan perbenua paling banyak dipegang oleh benua africa dengan total 61.45% dan benua dengan rata-rata paling sedikit di benua Europe total 18.77%


---

<br>
<a href="https://www.linkedin.com/in/retnosz/"><img src="https://img.shields.io/badge/-© 2024 Retno sari -417DAC?style=for-the-badge&logoColor=white"/></a>